### PR TITLE
[shared_preferences] Adds new `clearWithParameters` and `getAllWithParameters` methods to platform interface.

### PR DESCRIPTION
**Pull request by tarrinneal**
Adds new `clearWithParameters` and `getAllWithParameters` methods.

part of https://github.com/flutter/flutter/issues/128948

precursor to https://github.com/flutter/packages/pull/3794

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Review by stuartmorgan — APPROVED** [review]
LGTM with the shared_preferences changelog and version bump reverted.

File: packages/shared_preferences/shared_preferences/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+## 2.1.3
+
+* Ignores deprecation warnings.
```
[the remark this answers is not in the document]
Comment (tarrinneal) [reply packages/shared_preferences/shared_preferences/CHANGELOG.md:3]: I thought so, but the bot yelled at me.

[the remark this answers is not in the document]
Comment (stuartmorgan) [reply packages/shared_preferences/shared_preferences/CHANGELOG.md]: Right, that's where the override labels come in. The bot isn't looking at lines of diff yet, only changed files.

**Comment by stuartmorgan** [comment]
Overriding release metadata check: the changes in `shared_preferences` are dev-only.

**Comment by whesse** [comment]
I think this does need a version and a release, because the flutter_plugins tests check out these repositories, use pub_get to check out their dependencies, and then analyze them.

flutter_plugins is now failing on the main flutter CI, and on the monorepo CI.

**Comment by stuartmorgan** [comment]
> I think this does need a version and a release, because the flutter_plugins tests check out these repositories, use pub_get to check out their dependencies, and then analyze them.

I'm not following. Why would we need to publish for tests that are operating on copies of the plugins that are cloned from repo source?

> flutter_plugins is now failing on the main flutter CI, and on the monorepo CI.

Can you link to a failure? I would expect that to be the result of the pin not having rolled forward, rather than a lack of publishing.

**Comment by whesse** [comment]
The failing builder is

https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_plugins

a failing build is:

https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_plugins/12841/overview

Another failing builder, on the Dart CI, is 

https://ci.chromium.org/ui/p/dart/builders/ci.sandbox/flutter-linux-flutter-plugins

I think the failure may be due to a pinned package being chosen in the checkout, but then "flutter pub get" is run in that package to fetch its dependencies from the published versions.  That is why a package pinned in flutter still needs to be resolvable and correct against its publicly available dependencies.  That is what this flutter_plugins analysis test tests.

**Comment by stuartmorgan** [comment]
> The failing builder is
> 
> https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_plugins

That's definitely due to the pinned version in flutter/flutter not having rolled yet. It looks like the auto-roller is due to run in 20 minutes, or we could move the pin forward manually.

> Another failing builder, on the Dart CI, is
> 
> https://ci.chromium.org/ui/p/dart/builders/ci.sandbox/flutter-linux-flutter-plugins

I don't know how the Dart CI manages its pin, but that's also checking out a version of flutter/packages from before this PR landed.

> then "flutter pub get" is run in that package to fetch its dependencies from the published versions.

I'm aware of that, but we aren't running `analyze` against the `pub get`-fetched dependencies of a given package, we are analyzing the package itself, which is from source. If we published all of the higher-level packages right now, absolutely nothing would change about those failures.

> That is why a package pinned in flutter still needs to be resolvable and correct against its publicly available dependencies.

They are both resolvable and correct against public dependencies. They (and again, "they" here is *older versions of the packages that don't have the changes in this PR*) just have analysis warnings. The solution is to update the versions being analyzed to have the `// ignore` directives, which means rolling the flutter/packages checkout forward in those CI systems.

**Comment by whesse** [comment]
OK, I understand now.  I didn't see the difference between the pins of packages and of flutter_plugins, and that the package with the deprecations was being rolled separately from the package with the tests.  That is why I thought one of the two must be being fetched somehow.

**Comment by stuartmorgan** [comment]
> I didn't see the difference between the pins of packages and of flutter_plugins, and that the package with the deprecations was being rolled separately from the package with the tests.

I don't follow. There's just one pin involved here, not two separate rolls.

> That is why I thought one of the two must be being fetched somehow.

The package that introduced the deprecations (the one published in this PR) *is* being fetched via `pub get` in the context of the packages that are failing analysis. But that's not where the failures are, the failures are in the older copies of packages that are being analyzed from source.

**Pull request merged**